iceoryx_hoofs/test/moduletests/test_containers_uninitialized_array.cpp

```diff
@@ -0,0 +1,228 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "test.hpp"
+
+#include "iceoryx_hoofs/containers/uninitialized_array.hpp"
+
+namespace
+{
+using namespace ::testing;
+
+using iox::containers::UninitializedArray;
+
+template <typename T>
+class UninitializedArrayTest : public ::testing::Test
+{
+  public:
+    using Buffer = T;
+    Buffer buffer{};
+
+    void fillBuffer(int startValue)
+    {
+        auto capacity = buffer.capacity();
+
+        int value = startValue;
+        for (uint64_t i = 0; i < capacity; ++i)
+        {
+            buffer[i] = value++;
+        }
+    }
+};
+
+struct Integer
+{
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) required for typed tests
+    Integer(int value = 0)
+        : value(value)
+    {
+    }
+
+    int value{0};
+
+    // so that it behaves like an int for comparison purposes
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) required for typed tests
+    operator int() const
+    {
+        return value;
+    }
+};
+
+TEST(UninitializedArrayTest, capacityIsCorrect)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0ac31a08-77b2-4fd2-a214-81298cada00c");
+    constexpr uint64_t capacity = 7;
+    // NOLINTNEXTLINE(hicpp-member-init, cppcoreguidelines-pro-type-member-init) false positive
+    UninitializedArray<int, capacity> sut;
+    EXPECT_EQ(sut.capacity(), capacity);
+}
+
+typedef ::testing::Types<UninitializedArray<int, 10>,
+                         UninitializedArray<Integer, 10>,
+                         UninitializedArray<int, 10, iox::containers::FirstElementZeroed>,
+                         UninitializedArray<Integer, 10, iox::containers::FirstElementZeroed>>
+    TestArrays;
+
+TYPED_TEST_SUITE(UninitializedArrayTest, TestArrays, );
+
+TYPED_TEST(UninitializedArrayTest, accessElements)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f8621711-9c4c-45b5-979d-404357b664a4");
+    auto& buffer = this->buffer;
+    auto capacity = buffer.capacity();
+
+    int startValue = 73;
+    this->fillBuffer(startValue);
+
+    int expected = startValue;
+    for (uint64_t i = 0; i < capacity; ++i)
+    {
+        EXPECT_EQ(buffer[i], expected++);
+    }
+}
+
+TYPED_TEST(UninitializedArrayTest, accessElementsOfConstUinitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "50575bac-cff9-4896-89da-b03753370b18");
+    auto& buffer = this->buffer;
+    auto capacity = buffer.capacity();
+
+    int startValue = 37;
+    this->fillBuffer(startValue);
+
+    const auto& constBuffer = this->buffer;
+    int expected = startValue;
+    for (uint64_t i = 0; i < capacity; ++i)
+    {
+        EXPECT_EQ(constBuffer[i], expected++);
+    }
+}
+
+TEST(UninitializedArray, FirstElementIsInitializedWithZeroWhenBufferSetToFirstElementZeroed)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "bb213516-ab37-43e3-b2ec-098c98d777d1");
+    UninitializedArray<uint32_t, 2, iox::containers::FirstElementZeroed> buffer;
+    EXPECT_THAT(buffer[0], Eq(0));
```

Comment:
I think this is not enough. The reason is that most OS provide you initially zeroed memory for security reasons (so that one is unable to read the memory content of the previously running program). It is possible that you use most of the time fresh zeroed memory and even when this functionality does not work the test passes.

My suggestion would be to fill the memory first with garbage and then verify that the first element was zeroed and only the first element was zeroed!

```cxx
UninitializedArray<uint32_t, 2, iox::containers::FirstElementZeroed> buffer;
for( auto & e: buffer) {
  new (&e) uint32_t(std::numeric_limits<uint32_t>::max());
}

// now the memory is full with garbage and we construct via placement new a new Uninitialized
// Array at its position, if the policy works only the first element should be zero and the remaining
// elements should be uninitialized, e.g. the memory should be untouched.
new (&buffer) UninitializedArray<uint32_t, 2, iox::containers::FirstElementZeroed();

EXPECT_THAT(buffer[0], Eq(0));
for(auto &e : buffer) {
  EXPECT_THAT(e, Eq(std::numeric_limits<uint32_t>::max());
}
```

Reply:
The test revealed that everything is zeroed which makes sense according to https://en.cppreference.com/w/c/language/array_initialization, section 'Initialization from brace-enclosed lists'. I decided to keep the behavior and rename the policy parameters. Maybe we can add the first element zeroed policy later when needed.